src/translator/translation_model.cpp

```diff
@@ -162,11 +174,22 @@ Ptr<marian::data::CorpusBatch> TranslationModel::convertToMarianBatch(Batch &bat
   return corpusBatch;
 }
 
-void TranslationModel::translateBatch(size_t deviceId, Batch &batch) {
+void TranslationModel::translateBatch(Workspace &workspace, Batch &batch) {
+  size_t deviceId = workspace.id();
+
+  // Create backend if not exists, for device. Dynamically.
+  auto p = backend_.find(deviceId);
+  if (p == backend_.end()) {
+    backend_[deviceId] = MarianBackend{};
+    loadBackend(backend_[deviceId], workspace);
+  }
```

Comment:
> 3. How did swapping happen?

There is no swapping. Models are dynamically created here. Different tensors are created for different models and when the TranslationModel dies (and associated graph dies) the tensors are cleared. 

A backend is created if and only when the thread/workspace needs it for a translation. 

The graphs have shared TensorAllocator `tensors_` and `cache_`. There's an allocation-manager which owns and is aware of allocations tied to graph. When graph here, also scorer (loaded through `loadBackend`) dies with the TranslationModel hope is that the allocation-manager's destructer will free any remnant tensors and exit cleanly.